cartography/graph/loader.py

```diff
@@ -0,0 +1,144 @@
+import logging
+from typing import Dict
+from typing import List
+
+import neo4j
+
+from cartography.graph.querybuilder import build_node_ingestion_query
+from cartography.graph.querybuilder import build_relationship_ingestion_query
+
+logger = logging.getLogger(__name__)
+
+
+def merge_nodes(
+    neo4j_session: neo4j.Session, node_label: str, node_property_map: Dict[str, str],
+    node_data_list: List[Dict], update_tag: int, batch_size: int = 10000,
+) -> None:
+    """
+    Writes the `node_data_list` to Neo4j using the UNWIND pattern in batches with
+    just one MERGE per statement for performance and reliability. Follows guidelines
+    described in
+    - https://neo4j.com/blog/bulk-data-import-neo4j-3-0/
+    - https://dzone.com/articles/tips-for-fast-batch-updates-of-graph-structures-wi
+
+    The query run looks like
+
+    UNWIND {DictList} AS item
+        MERGE (i:`node_label`{id:item.`node_property_map['id']`})
+        ON CREATE SET i.firstseen = timestamp()
+        SET i.lastupdated = {UpdateTag},
+        ... <expand the given node property map to set the other node fields> ...
+
+    :param neo4j_session: The Neo4j session
+    :param node_label: The label of the node to MERGE to Neo4j
+    :param node_property_map: Mapping from node field names to dict keys in `node_data_list`.
+    :param node_data_list: A list of dicts to be written as nodes to the graph.
```

Comment:
can we simplify these three parameters? in particular the `node_property_map` feels inelegant, mapping Neo4j field names to dict field names. I'd rather just have the dict field names be the same as Neo4j field names.

An idea in the same vein: instead of having all of these parameters be separate entities, what if this method just took a list of dataclass objects? The dataclass would have a name (`node_label`), named fields (replacing the `node_property_map`), and data for each of those fields (`node_data_list`). The field names in the dataclass could be the exact same names attached to the Neo4j objects. Dataclasses are pretty simple and elegant and are typically as efficient as dicts but with better type checks, syntax highlighting, and readability

Reply:
> in particular the node_property_map feels inelegant, mapping Neo4j field names to dict field names. I'd rather just have the dict field names be the same as Neo4j field names.

Makes sense, although the `transform` step of every intelmodule would now need to change the key of each dictionary to be lowercase.

On dataclasses, I'm a fan. Before adding that functionality I wanted to share this PR first to show what is going on under the hood to get a better idea of how we would map things together. How would we represent relationships in dataclasses though? This current approach keeps that part fairly lightweight with just lists of dicts for nodes, and lists of dicts representing node-mapping pairs for relationships.